File: sdk/python/unagnt/stream.py
"""Streaming support for Agent Runtime Python SDK"""

import asyncio
from typing import AsyncIterator, Dict, Any
import httpx

from .types import Run
from .errors import AgentRuntimeError, APIError


class StreamEvent:
    """Represents a streaming event from the agent runtime"""
    
    def __init__(self, data: Dict[str, Any]):
        self.run_id = data.get("run_id")
        self.step_id = data.get("step_id")
        self.timestamp = data.get("timestamp")
        self.type = data.get("type")
        self.agent = data.get("agent")
        self.data = data.get("data", {})
        self.model = data.get("model", {})
    
    def __repr__(self):
        return f"StreamEvent(type={self.type}, run_id={self.run_id})"


async def stream_events(
    base_url: str,
    run_id: str,
    api_key: str = None,
    timeout: float = 300.0
) -> AsyncIterator[StreamEvent]:
    """
    Stream events for a run using Server-Sent Events.
    
    Args:
        base_url: Base URL of the agent runtime API
        run_id: The run ID to stream events for
        api_key: Optional API key for authentication
        timeout: Request timeout in seconds
    
    Yields:
        StreamEvent objects as they arrive
    
    Example:
        async for event in stream_events("http://localhost:8080", run_id):
            print(f"Event: {event.type}")
    """
    url = f"{base_url.rstrip('/')}/v1/runs/{run_id}/stream"
    
    headers = {
        "Accept": "text/event-stream",
    }
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    
    async with httpx.AsyncClient(timeout=timeout) as client:
        try:
            async with client.stream("GET", url, headers=headers) as response:
                if response.status_code != 200:
                    raise APIError(
                        response.status_code,
                        f"Stream failed: {response.text}"
                    )
                
                async for line in response.aiter_lines():
                    # Skip empty lines and comments (heartbeats)
                    if not line or line.startswith(":"):
                        continue
                    
                    # SSE format: "data: {...}"
                    if line.startswith("data: "):
                        data_str = line[6:]
                        
                        try:
                            import json
                            data = json.loads(data_str)
                            yield StreamEvent(data)
                        except json.JSONDecodeError as e:
                            raise AgentRuntimeError(f"Failed to parse event: {e}")
                        
        except httpx.TimeoutException as e:
            raise AgentRuntimeError(f"Stream timeout: {e}")
        except httpx.RequestError as e:
            raise AgentRuntimeError(f"Stream request failed: {e}")

File: sdk/go/client/stream.go
package client

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// Event represents a streaming event from the agent runtime.
type Event struct {
	RunID     string         `json:"run_id"`
	StepID    string         `json:"step_id,omitempty"`
	Timestamp time.Time      `json:"timestamp"`
	Type      string         `json:"type"`
	Agent     string         `json:"agent"`
	Data      map[string]any `json:"data,omitempty"`
	Model     ModelMeta      `json:"model,omitempty"`
}

// ModelMeta captures model metadata.
type ModelMeta struct {
	Provider string `json:"provider,omitempty"`
	Name     string `json:"name,omitempty"`
	Version  string `json:"version,omitempty"`
}

// StreamEvents opens a Server-Sent Events connection to stream run events.
func (c *Client) StreamEvents(ctx context.Context, runID string) (<-chan Event, <-chan error, error) {
	url := fmt.Sprintf("%s/v1/runs/%s/stream", c.baseURL, runID)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, nil, err
	}

	if c.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.apiKey)
	}
	req.Header.Set("Accept", "text/event-stream")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, nil, err
	}

	if resp.StatusCode != http.StatusOK {
		resp.Body.Close()
		return nil, nil, fmt.Errorf("stream failed: status %d", resp.StatusCode)
	}

	eventChan := make(chan Event, 10)
	errChan := make(chan error, 1)

	go func() {
		defer close(eventChan)
		defer close(errChan)
		defer resp.Body.Close()

		scanner := bufio.NewScanner(resp.Body)

		for scanner.Scan() {
			line := scanner.Text()

			// Skip empty lines and comments (heartbeats)
			if line == "" || line[0] == ':' {
				continue
			}

			// SSE format: "data: {...}"
			if len(line) > 6 && line[:6] == "data: " {
				data := line[6:]

				var event Event
				if err := json.Unmarshal([]byte(data), &event); err != nil {
					errChan <- fmt.Errorf("parse event: %w", err)
					return
				}

				select {
				case eventChan <- event:
				case <-ctx.Done():
					return
				}
			}
		}

		if err := scanner.Err(); err != nil {
			select {
			case errChan <- fmt.Errorf("scan error: %w", err):
			case <-ctx.Done():
			}
		}
	}()

	return eventChan, errChan, nil
}
